discojs/src/client/client.ts

```diff
@@ -45,8 +69,96 @@ export abstract class Client extends EventEmitter<{'status': RoundStatus}>{
   /**
    * Handles the disconnection process of the client from any sort of network server.
    */
-  async disconnect (): Promise<void> {}
+  async disconnect(): Promise<void> { }
+
+  /**
+   * Emits the round status specified. It also stores the status emitted such that 
+   * if the server tells the client to wait for more participants, it can display
+   * the waiting status and once enough participants join, it can display the previous status again
+   */ 
+  protected saveAndEmit(status: RoundStatus) {
+    this.previousStatus = status
+    this.emit("status", status)
+  }
+  
+  /**
+   * For both federated and decentralized clients, we listen to the server to tell
+   * us whether there are enough participants to train. If not, we pause until further notice.
+   * When a client connects to the server, the server answers with the session information (id, 
+   * number of participants) and whether there are enough participants. 
+   * When there are the server sends a new EnoughParticipant message to update the client. 
+   * 
+   * `setMessageInversionFlag` is used to address the following scenario:
+   * 1. Client 1 connect to the server
+   * 2. Server answers with message A containing "not enough participants"
+   * 3. Before A arrives a new client joins. There are enough participants now.
+   * 4. Server updates client 1 with message B saying "there are enough participants"
+   * 5. Due to network and message sizes message B can arrive before A. 
+   * i.e. "there are enough participants" arrives before "not enough participants"
+   * ending up with client 1 thinking it needs to wait for more participants.
+   * 
+   * To keep track of this message inversion, `setMessageInversionFlag` 
+   * tells us whether a message inversion occurred (by setting a boolean to true)
```

Comment:
I completely agree, do you know of paradigms/libraries/documentation on how to manage that? (I'll research it myself, just in case you already have some knowledge on the subject)

Reply:
I've been reflecting a while on how to best approach it. I'm still unsure, one would need to validate it with some code to know how well it works, but it would already help us see more clearly:

let's first remark that many parts requires a different set of message (federated vs decentralized, secure aggregator vs some other) yet are still used in the same fashion. so it seems that we can plug a class with various implementations.

we can have an abstract `NetworkClient` with various implementation, one federated, one decentralized (and one local). theses advertise their supported messages by exposing a `isValidMessage(raw: unknown): raw is Message` (with `Message` depending on the implementation). that ensure that the type of messages we are receiving are typed correctly. it then gets passed an async queue of messages (such as [`reactive-channel`](https://www.npmjs.com/package/reactive-channel)'s one).
same idea for `AggregatorClient`

`Client` should do more, and not be an abstract class but be constructed with `NetworkClient`, `AggregatorClient`, …. it should take care of splitting the message stream, ie read next msg and put it the queue where the corresponding `isValidMessage` succeeded.

that helps by having a local set of messages per implementation, and only care about these ones. we can also better check that the server is not weirdly behaving by failing when the next messages is unexpected. and message order is preserved because async channels are also ordered and there is a single reader of WebSocket.

with that, it forces us to check that we don't receive an unexpected message in the stream. that can be a bit bothering in the long run (I've other ways to fix that such as type-encoding a state-transition table, but I might be going a bit far for now). but overall, I do think that can greatly help us understand where messages are coming from and in which state we are.
WDYT for a first draft?

Reply:
Yes that sounds good! Already using a queue would ensure sending messages in order and using different message channels makes things extra clear. And the server would also need to rely on the same message queues I think. Do you mind creating an issue to keep track of this?